app.py

```diff
@@ -1,31 +1,307 @@
 
 app = Flask(__name__)
+app.secret_key = 'some_secret'
 
-@app.route('/')
-def hello():
-    return 'Hello World'
+app.config["MONGO_DBNAME"] = 'cookbookdatabase'
+app.config["MONGO_URI"] ="mongodb://admin:Komplete8@ds119150.mlab.com:19150/cookbookdatabase"
+
+mongo = PyMongo(app)
+
+country_list = get_countries()
+
+class ReusableForm(Form):
+    
+    #Set up form
+    name = TextField('Recipe Name:', validators=[validators.DataRequired("*Required")])
+    description = TextField('Description:', validators=[validators.DataRequired("*Required")])
+    author = TextField('Author:', validators=[validators.DataRequired("*Required")])
+    instruction1 = TextField('Step 1:', validators=[validators.DataRequired("*Required")])
+    ingredient1 = TextField( validators=[validators.DataRequired("*Required")])
+    country = SelectField('Country of Origin', choices=country_list , validators=[validators.DataRequired()])
+
+class Username(Form):
+    #Set up form
+    username = TextField('Username:', validators=[validators.DataRequired("*Required")])
+
+class Search(Form):
+    #Set up form
+    search = TextField('Search:', validators=[validators.DataRequired("*Required")])
+
+@app.route('/', methods=['GET','POST'])
+def index():
+    #Username form
+    wtform = Username(request.form)
+    if wtform.validate():
+        return redirect('/' + request.form['username'] + '/recipes?limit=10&offset=0')
+    
+    return render_template("index.html", form=wtform, errors=wtform.errors)
+    
+@app.route('/<username>/recipes')
+def recipes(username):
+    
+    # Get All Recipes
+    recipes = mongo.db.recipes
+    all_recipes = recipes.find()
+    
+    # Pagination Settings
+    offset = 0 
+    limit = int(request.args.get('limit'))
+    
+    #Get Count
+    count_list = []
+    for doc in all_recipes:
+        count_list.append(doc)
+        count = len(count_list)
+    
+    #Get Pages And Generate URL List
+    pages = get_pages(count, limit)
+    url_list = generate_pagination_links(offset, limit, pages, 'recipes', 'null', username)
+    print(url_list)
+    
+    #Get _id of Last Item on a Page
+    dynamic_position = request.args.get('offset')
+    starting_id = recipes.find().sort('_id')
+    last_id = starting_id[int(dynamic_position)]['_id']
+    
+    #Sort Tables
+    sort_default = recipes.find({'_id' : {'$gte' : last_id}}).limit(limit)
+    sort_country = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "country", 1),
+    ("name", 1 )]).limit(limit)
+    sort_name = recipes.find({'_id' : {'$gte' : last_id}}).sort([("name",
+    1 )]).limit(limit)
+    sort_upvotes = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "upvotes",
+    pymongo.DESCENDING), ("name",1 )]).limit(limit)
+    sort_downvotes = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "downvotes",pymongo.DESCENDING), ("name", 1 )]).limit(limit)
+    sort_author = recipes.find({'_id' : {'$gte' : last_id}}).sort([( "author",
+    1), ("name", 1 )]).limit(limit)
```

Comment:
Read up on MVC architecture to understand the concept of model, views and controllers. Then create a new file and call it `controllers.py` where you will have the logic of your CRUD functions, like what you have on page  53 - 86. Then all you need to do here is import the file and call the function passing in relevant arguments. This will keep you code clean and declarative. Same goes for other CRUD logic.

Reply:
Spent some hours trying to get this working and didn't have any luck. It's the same issue I encountered earlier in the development that I described on our last call.